### ISSUE TITLE
How do I get the output of DLRM to support multiple classes?

### ISSUE THREAD

**Issue by danisben** [facebookresearch/dlrm#17]
Hi, DLRM can be used for binary output (0 or 1) currently, and it works well. 
But I'm not able to change it for multiple classes support. Do you know how can this be done? Thx!

**Comment by mnaumovfb** [comment]
I think the simplest approach is to switch from i) binary cross entropy to ii) cross entropy loss 
i)  https://pytorch.org/docs/stable/nn.html#binary-cross-entropy 
ii) https://pytorch.org/docs/stable/nn.html#cross-entropy
and adjust the targets accordingly (and/or work with your own dataset with muti-class targets).

**Comment by danisben** [comment]
@mnaumovfb Thanks for the reply! 

In the original case, when the label of my data was two classes, the accuracy was 77.8%, which was a good result. But when I adjust the label of my data into 5 categories, the accuracy becomes 134%, which is a little strange. (the datasets of two cases with same features and data)

My modifications to the code are as follows:
1) switch the loss function from "BCELoss" to "CrossEntropyLoss"
2) adjust the top MLP from "512-256-1" to "512-256-5" by changing the "--arch-mlp-top" parameter, which indicates 5 classes.
3) In the loss_fn_wrap function, switch "loss_fn(Z, T)" to "loss_fn(Z, T.long().squeeze(1))", in order to adapt the "CrossEntropyLoss" loss function.

**Comment by mnaumovfb** [comment]
The steps you have described seem reasonable to me. However, the accuracy of 134% does not make sense. I would carefully check your code for bugs. 

Notice that in the original version a single output "p" represents both categories, in other words, they are not represented by two outputs [0,1] and [1,0]. I suspect your bug is related to this change (except you have 5 categories). Try to print and see whether the following line which computes accuracy is still doing what you expect after your changes:
https://github.com/facebookresearch/dlrm/blob/master/dlrm_s_pytorch.py#L756

**Comment by danisben** [comment]
I think you are correct. I printed out the "A" variable in Line 756, it showed the list of values that around 1.34, indicates the average 134%.
![image](https://user-images.githubusercontent.com/16700708/61753493-8bd4b100-ad7d-11e9-9985-1198ccb0bc40.png)

Then I checked the output of the network "S", it showed the arrays contains 5 values, representing the 5 categories.
![image](https://user-images.githubusercontent.com/16700708/61753572-d0604c80-ad7d-11e9-86b2-5cf0cf10acf2.png)

And the variable T shows the Target (values between 0 to 4 representing the 5 categories):
![image](https://user-images.githubusercontent.com/16700708/61753624-056c9f00-ad7e-11e9-93e7-1fa784185d64.png)

So the reason is related to the variable "S", here is the "S" from the case which contains two categories (0 and 1):
![image](https://user-images.githubusercontent.com/16700708/61753800-a6f3f080-ad7e-11e9-94fc-35bf307803b6.png)

So I think the precision calculation of line 765 can not adapt to multi-category judgment. How can we optimize for this requirement? Thx.

**Comment by mnaumovfb** [comment]
I think that you should be able to relatively easily adjust the accuracy calculation for your use case. Just make sure that you calculate the number of correctly predicted samples based on your output.

**Comment by joy80** [comment]
`python dlrm_s_pytorch.py --data-generation=dataset --data-set=kaggle --raw-data-file=train.txt --loss-function=bce --round-targets=True --learning-rate=0.1 --mini-batch-size=128 --test-freq=10 --save-model=model.pt`

after saving the model by this command how can I get the output from the trained model?
`python dlrm_s_pytorch.py --load-model=model.pt --data-generation=dataset --data-set=kaggle --raw-data-file=test.txt`
is the command is correct for the output purpose?

dlrm_s_pytorch.py line=275
 `z = p`
is z is the outputs?
please help !

**Comment by danisben** [comment]
Thank you so much. I will recalculate the accuracy based on my case.

And I have two more questions:

1) I would like to ask whether, for all cases, the amount of data needs to be more than 10,000 to get a reasonable result? When I used the Kaggle Display Advertising Challenge Dataset for testing, I found that when the batch size was 128, only after about 100 batches of training could top MLP output the predicted value greater than 0.5.

2) How to define the batch size? Will it have any impact on the result?

**Comment by mnaumovfb** [comment]
1. I would not rely on a particular portion of data for training. The progression of training will depend on the dataset and the ordering of samples in it, among other things. 

2. The mini-batch size does affect the convergence of different optimization methods during training, see [1] and follow up work on it. I would experiment with it and set it empirically based on your runs.  

[1] https://arxiv.org/abs/1609.04836v1

**Comment by mnaumovfb** [comment]
It looks like this issue is resolved. Closing.
